widgyts/dataset_viewer.py

```diff
@@ -0,0 +1,252 @@
+import numpy as np
+import json
+import pythreejs
+import traitlets
+import ipywidgets
+from IPython.display import display, JSON
+import matplotlib.cm as mcm
+import matplotlib.colors as mcolors
+from yt.data_objects.api import Dataset
+from yt.units.yt_array import display_ytarray
+
+
+class DatasetViewerComponent(traitlets.HasTraits):
+    ds = traitlets.Instance(Dataset)
+    viewer = traitlets.ForwardDeclaredInstance("DatasetViewer")
+
+
+class DatasetViewer(traitlets.HasTraits):
+    ds = traitlets.Instance(Dataset)
+    components = traitlets.List(trait=traitlets.Instance(DatasetViewerComponent))
+
+    @traitlets.default("components")
+    def _components_default(self):
+        adv = AMRDomainViewer(ds=self.ds, viewer=self)
+        fdv = FieldDefinitionViewer(ds=self.ds, viewer=self)
+        pv = ParametersViewer(ds=self.ds, viewer=self)
+        return [adv, fdv, pv]
+
+    def widget(self):
+        tab = ipywidgets.Tab(
+            children=[_.widget() for _ in self.components],
+            layout=ipywidgets.Layout(height="500px"),
+        )
+        for i, c in enumerate(self.components):
+            tab.set_title(i, c.name)
+        return tab
+
+
+class FieldDefinitionViewer(DatasetViewerComponent):
+    name = "Fields"
+
+    def widget(self):
+        out = ipywidgets.Output()
+        with out:
+            display(self.ds.fields)
+        return out
+
+
+class DomainViewer(DatasetViewerComponent):
+    domain_axes = traitlets.Instance(pythreejs.AxesHelper)
+    name = "Domain Viewer"
+
+    @traitlets.default("domain_axes")
+    def _domain_axes_default(self):
+        offset_vector = (self.ds.domain_left_edge - self.ds.domain_center) * 0.1
+        position = tuple(
+            (self.ds.domain_left_edge + offset_vector).in_units("code_length").d
+        )
+        # We probably don't want to use the AxesHelper as it doesn't expose the
+        # material, which can result in it not being easy to see.  But for now...
+        ah = pythreejs.AxesHelper(
+            position=position,
+            scale=tuple(self.ds.domain_width.in_units("code_length").d),
+        )
+        return ah
+
+
+_CORNER_INDICES = np.array(
```

Comment:
Is there a plan to use this elsewhere? Or should it be moved ~l94 or so?

Reply:
I anticipate that if we have annotations, etc, they could use this, but I'm also hesitant to do a bunch of static declarations inside the function body -- I thought it would be clearer here.

Reply:
Ok, if we plan on having other annotations that use it then I think it should be moved up to after imports, rather than between classes.